### DISCUSSION TITLE
训练时，ppocr的打印未出现eval，只是一直save model并且最后打印best metric, hmean: 0, is_float16: False

### DISCUSSION THREAD

**Discussion by Jasonchanlan** [PaddlePaddle/PaddleOCR#14951] [Q&A]
python tools/train.py -c configs/det/ch_PP-OCRv4/ch_PP-OCRv4_det_student.yml
信息: 用提供的模式无法找到文件。
E:\anconda\envs\paddleOCR\lib\site-packages\paddle\utils\cpp_extension\extension_utils.py:711: UserWarning: No ccache found. Please be aware that recompiling all source files may be required. You can download and install ccache from: https://github.com/ccache/ccache/blob/master/doc/INSTALL.md
  warnings.warn(warning_message)
[2025/03/28 17:45:00] ppocr WARNING: Skipping import of the encryption module.
[2025/03/28 17:45:00] ppocr INFO: Architecture :
[2025/03/28 17:45:00] ppocr INFO:     Backbone :
[2025/03/28 17:45:00] ppocr INFO:         det : True
[2025/03/28 17:45:00] ppocr INFO:         name : PPLCNetV3
[2025/03/28 17:45:00] ppocr INFO:         scale : 0.75
[2025/03/28 17:45:00] ppocr INFO:     Head :
[2025/03/28 17:45:00] ppocr INFO:         fix_nan : True
[2025/03/28 17:45:00] ppocr INFO:         k : 50
[2025/03/28 17:45:00] ppocr INFO:         name : DBHead
[2025/03/28 17:45:00] ppocr INFO:     Neck :
[2025/03/28 17:45:00] ppocr INFO:         name : RSEFPN
[2025/03/28 17:45:00] ppocr INFO:         out_channels : 96
[2025/03/28 17:45:00] ppocr INFO:         shortcut : True
[2025/03/28 17:45:00] ppocr INFO:     Transform : None
[2025/03/28 17:45:00] ppocr INFO:     algorithm : DB
[2025/03/28 17:45:00] ppocr INFO:     model_type : det
[2025/03/28 17:45:00] ppocr INFO: Eval :
[2025/03/28 17:45:00] ppocr INFO:     dataset :
[2025/03/28 17:45:00] ppocr INFO:         data_dir : ./train_data/
[2025/03/28 17:45:00] ppocr INFO:         label_file_list : ['./train_data/det/val.txt']
[2025/03/28 17:45:00] ppocr INFO:         name : SimpleDataSet
[2025/03/28 17:45:00] ppocr INFO:         transforms :
[2025/03/28 17:45:00] ppocr INFO:             DecodeImage :
[2025/03/28 17:45:00] ppocr INFO:                 channel_first : False
[2025/03/28 17:45:00] ppocr INFO:                 img_mode : BGR
[2025/03/28 17:45:00] ppocr INFO:             DetLabelEncode : None
[2025/03/28 17:45:00] ppocr INFO:             DetResizeForTest : None
[2025/03/28 17:45:00] ppocr INFO:             NormalizeImage :
[2025/03/28 17:45:00] ppocr INFO:                 mean : [0.485, 0.456, 0.406]
[2025/03/28 17:45:00] ppocr INFO:                 order : hwc
[2025/03/28 17:45:00] ppocr INFO:                 scale : 1./255.
[2025/03/28 17:45:00] ppocr INFO:                 std : [0.229, 0.224, 0.225]
[2025/03/28 17:45:00] ppocr INFO:             ToCHWImage : None
[2025/03/28 17:45:00] ppocr INFO:             KeepKeys :
[2025/03/28 17:45:00] ppocr INFO:                 keep_keys : ['image', 'shape', 'polys', 'ignore_tags']
[2025/03/28 17:45:00] ppocr INFO:     loader :
[2025/03/28 17:45:00] ppocr INFO:         batch_size_per_card : 1
[2025/03/28 17:45:00] ppocr INFO:         drop_last : False
[2025/03/28 17:45:00] ppocr INFO:         num_workers : 2
[2025/03/28 17:45:00] ppocr INFO:         shuffle : False
[2025/03/28 17:45:00] ppocr INFO: Global :
[2025/03/28 17:45:00] ppocr INFO:     cal_metric_during_train : True
[2025/03/28 17:45:00] ppocr INFO:     checkpoints : None
[2025/03/28 17:45:00] ppocr INFO:     d2s_train_image_shape : [3, 640, 640]
[2025/03/28 17:45:00] ppocr INFO:     debug : True
[2025/03/28 17:45:00] ppocr INFO:     distributed : False
[2025/03/28 17:45:00] ppocr INFO:     epoch_num : 10
[2025/03/28 17:45:00] ppocr INFO:     eval_batch_step : [0, 5]
[2025/03/28 17:45:00] ppocr INFO:     infer_img : doc/imgs_en/img_10.jpg
[2025/03/28 17:45:00] ppocr INFO:     log_smooth_window : 20
[2025/03/28 17:45:00] ppocr INFO:     pretrained_model : ./pretrain_models/ch_PP-OCRv4_det_train/best_accuracy
[2025/03/28 17:45:00] ppocr INFO:     print_batch_step : 2
[2025/03/28 17:45:00] ppocr INFO:     save_epoch_step : 2
[2025/03/28 17:45:00] ppocr INFO:     save_inference_dir : None
[2025/03/28 17:45:00] ppocr INFO:     save_model_dir : ./output/ch_PP-OCRv4
[2025/03/28 17:45:00] ppocr INFO:     save_res_path : ./checkpoints/det_db/predicts_db.txt
[2025/03/28 17:45:00] ppocr INFO:     use_gpu : False
[2025/03/28 17:45:00] ppocr INFO:     use_visualdl : False
[2025/03/28 17:45:00] ppocr INFO: Loss :
[2025/03/28 17:45:00] ppocr INFO:     alpha : 5
[2025/03/28 17:45:00] ppocr INFO:     balance_loss : True
[2025/03/28 17:45:00] ppocr INFO:     beta : 10
[2025/03/28 17:45:00] ppocr INFO:     main_loss_type : DiceLoss
[2025/03/28 17:45:00] ppocr INFO:     name : DBLoss
[2025/03/28 17:45:00] ppocr INFO:     ohem_ratio : 3
[2025/03/28 17:45:00] ppocr INFO: Metric :
[2025/03/28 17:45:00] ppocr INFO:     main_indicator : hmean
[2025/03/28 17:45:00] ppocr INFO:     name : DetMetric
[2025/03/28 17:45:00] ppocr INFO: Optimizer :
[2025/03/28 17:45:00] ppocr INFO:     beta1 : 0.9
[2025/03/28 17:45:00] ppocr INFO:     beta2 : 0.999
[2025/03/28 17:45:00] ppocr INFO:     lr :
[2025/03/28 17:45:00] ppocr INFO:         learning_rate : 0.001
[2025/03/28 17:45:00] ppocr INFO:         name : Cosine
[2025/03/28 17:45:00] ppocr INFO:         warmup_epoch : 2
[2025/03/28 17:45:00] ppocr INFO:     name : Adam
[2025/03/28 17:45:00] ppocr INFO:     regularizer :
[2025/03/28 17:45:00] ppocr INFO:         factor : 5e-05
[2025/03/28 17:45:00] ppocr INFO:         name : L2
[2025/03/28 17:45:00] ppocr INFO: PostProcess :
[2025/03/28 17:45:00] ppocr INFO:     box_thresh : 0.6
[2025/03/28 17:45:00] ppocr INFO:     max_candidates : 1000
[2025/03/28 17:45:00] ppocr INFO:     name : DBPostProcess
[2025/03/28 17:45:00] ppocr INFO:     thresh : 0.3
[2025/03/28 17:45:00] ppocr INFO:     unclip_ratio : 1.5
[2025/03/28 17:45:00] ppocr INFO: Train :
[2025/03/28 17:45:00] ppocr INFO:     dataset :
[2025/03/28 17:45:00] ppocr INFO:         data_dir : ./train_data/
[2025/03/28 17:45:00] ppocr INFO:         label_file_list : ['./train_data/det/train.txt']
[2025/03/28 17:45:00] ppocr INFO:         name : SimpleDataSet
[2025/03/28 17:45:00] ppocr INFO:         ratio_list : [1.0]
[2025/03/28 17:45:00] ppocr INFO:         transforms :
[2025/03/28 17:45:00] ppocr INFO:             DecodeImage :
[2025/03/28 17:45:00] ppocr INFO:                 channel_first : False
[2025/03/28 17:45:00] ppocr INFO:                 img_mode : BGR
[2025/03/28 17:45:00] ppocr INFO:             DetLabelEncode : None
[2025/03/28 17:45:00] ppocr INFO:             CopyPaste : None
[2025/03/28 17:45:00] ppocr INFO:             IaaAugment :
[2025/03/28 17:45:00] ppocr INFO:                 augmenter_args :
[2025/03/28 17:45:00] ppocr INFO:                     args :
[2025/03/28 17:45:00] ppocr INFO:                         p : 0.5
[2025/03/28 17:45:00] ppocr INFO:                     type : Fliplr
[2025/03/28 17:45:00] ppocr INFO:                     args :
[2025/03/28 17:45:00] ppocr INFO:                         rotate : [-10, 10]
[2025/03/28 17:45:00] ppocr INFO:                     type : Affine
[2025/03/28 17:45:00] ppocr INFO:                     args :
[2025/03/28 17:45:00] ppocr INFO:                         size : [0.5, 3]
[2025/03/28 17:45:00] ppocr INFO:                     type : Resize
[2025/03/28 17:45:00] ppocr INFO:             EastRandomCropData :
[2025/03/28 17:45:00] ppocr INFO:                 keep_ratio : True
[2025/03/28 17:45:00] ppocr INFO:                 max_tries : 50
[2025/03/28 17:45:00] ppocr INFO:                 size : [640, 640]
[2025/03/28 17:45:00] ppocr INFO:             MakeBorderMap :
[2025/03/28 17:45:00] ppocr INFO:                 shrink_ratio : 0.4
[2025/03/28 17:45:00] ppocr INFO:                 thresh_max : 0.7
[2025/03/28 17:45:00] ppocr INFO:                 thresh_min : 0.3
[2025/03/28 17:45:00] ppocr INFO:                 total_epoch : 10
[2025/03/28 17:45:00] ppocr INFO:             MakeShrinkMap :
[2025/03/28 17:45:00] ppocr INFO:                 min_text_size : 8
[2025/03/28 17:45:00] ppocr INFO:                 shrink_ratio : 0.4
[2025/03/28 17:45:00] ppocr INFO:                 total_epoch : 10
[2025/03/28 17:45:00] ppocr INFO:             NormalizeImage :
[2025/03/28 17:45:00] ppocr INFO:                 mean : [0.485, 0.456, 0.406]
[2025/03/28 17:45:00] ppocr INFO:                 order : hwc
[2025/03/28 17:45:00] ppocr INFO:                 scale : 1./255.
[2025/03/28 17:45:00] ppocr INFO:                 std : [0.229, 0.224, 0.225]
[2025/03/28 17:45:00] ppocr INFO:             ToCHWImage : None
[2025/03/28 17:45:00] ppocr INFO:             KeepKeys :
[2025/03/28 17:45:00] ppocr INFO:                 keep_keys : ['image', 'threshold_map', 'threshold_mask', 'shrink_map', 'shrink_mask']
[2025/03/28 17:45:00] ppocr INFO:     loader :
[2025/03/28 17:45:00] ppocr INFO:         batch_size_per_card : 8
[2025/03/28 17:45:00] ppocr INFO:         drop_last : False
[2025/03/28 17:45:00] ppocr INFO:         num_workers : 8
[2025/03/28 17:45:00] ppocr INFO:         shuffle : True
[2025/03/28 17:45:00] ppocr INFO: profiler_options : None
[2025/03/28 17:45:00] ppocr INFO: train with paddle 3.0.0-rc1 and device Place(cpu)
[2025/03/28 17:45:00] ppocr INFO: Initialize indexes of datasets:['./train_data/det/train.txt']
[2025/03/28 17:45:00] ppocr INFO: Initialize indexes of datasets:['./train_data/det/val.txt']
[2025/03/28 17:45:00] ppocr INFO: train dataloader has 1 iters
[2025/03/28 17:45:00] ppocr INFO: valid dataloader has 2 iters
[2025/03/28 17:45:00] ppocr WARNING: The pretrained params backbone.conv1.hardswish.scale not in model
[2025/03/28 17:45:00] ppocr WARNING: The pretrained params backbone.conv1.hardswish.bias not in model
[2025/03/28 17:45:00] ppocr INFO: load pretrain successful from ./pretrain_models/ch_PP-OCRv4_det_train/best_accuracy
[2025/03/28 17:45:00] ppocr INFO: During the training process, after the 0th iteration, an evaluation is run every 5 iterations
[2025/03/28 17:45:01] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:02] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:02] ppocr INFO: save model in ./output/ch_PP-OCRv4\iter_epoch_2
[2025/03/28 17:45:03] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:04] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:04] ppocr INFO: save model in ./output/ch_PP-OCRv4\iter_epoch_4
[2025/03/28 17:45:05] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:06] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:06] ppocr INFO: save model in ./output/ch_PP-OCRv4\iter_epoch_6
[2025/03/28 17:45:06] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:07] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:07] ppocr INFO: save model in ./output/ch_PP-OCRv4\iter_epoch_8
[2025/03/28 17:45:08] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:09] ppocr INFO: save model in ./output/ch_PP-OCRv4\latest
[2025/03/28 17:45:09] ppocr INFO: save model in ./output/ch_PP-OCRv4\iter_epoch_10
[2025/03/28 17:45:09] ppocr INFO: best metric, hmean: 0, is_float16: False

**Comment by GreatV** [comment]
从你提供的训练日志和配置来看，训练过程中的确并没有明显进行eval（评估）的相关打印输出，最终best metric为 hmean: 0，说明整个训练过程中模型在验证集上的检测指标为0，可能的原因如下：

1. ⚠️验证集数据标注错误或者无数据有效载入：
训练日志中提到：
[2025/03/28 17:45:00] ppocr INFO: valid dataloader has 2 iters

意味着验证集只有两条数据，数量非常少，考虑以下问题：

- 确保路径 ./train_data/det/val.txt 中的图像路径是否正确？
- 检查图像是否实际存在？
- 检查标注格式是否和 PaddleOCR 要求一致（例如：格式为 image_path\t[polygon list]\n）；
- 图像中多边形标注可能被认为是被忽略区域（如全为 ignore_tags: true），则对评估无贡献；

2. ☑️评估机制配置并未生效或被跳过：
配置中包含以下信息：
[2025/03/28 17:45:00] ppocr INFO: eval_batch_step : [0, 5]，代表每5步进行一次评估（从第0步开始），但后续日志并未见有“Hmean”或“precision/recall”等打印。这种现象可能因为以下原因：

- 数据集中加载的数据为0或加载失败，导致实际没有评估可进行；
- 设置了 debug: True，导致部分代码路径更简化，仅输出 minimal 日志对调试友好，建议将 debug 改为 False 试试看。

3. 📉模型输出为空（或预测框全为空）导致评估指标为 0：
即便验证集数据读取成功，如果模型在验证集上的预测输出为空，则 hmean 会为0。这可能由于：

- 学习率过低或过高，模型未有效学习；
- 数据量太少（如你的 val.txt 仅有两条图片），模型无法泛化；
- 模型预训练参数加载不匹配，从日志可见：

[2025/03/28 17:45:00] ppocr WARNING: The pretrained params backbone.conv1.hardswish.scale not in model

这说明你加载的预训练模型和当前网络结构可能不完全一致，可能导致模型初始化不好，影响性能；

4. ✅参考解决建议：

- 确保验证集数据格式正确，标注没有问题（推荐手动检查 ./train_data/det/val.txt 和对应图片）；
- 增加验证集数据规模；
- 关闭 debug 模式，确认是否打印出 val 的预测信息；
- 检查训练集是否足够、是否过拟合（如训练很快 loss 收敛，但 val 无预测）；
- 确保模型结构和预训练参数完全匹配；
- 如有必要，可将 cal_metric_during_train 设置为 False，改为单独评估模型（tools/eval.py）以更可控地调试验证集效果；

你可以参考下相关讨论：https://github.com/PaddlePaddle/PaddleOCR/discussions/13839，其中也有类似关于 hmean: 0 的问题。

如果进一步检查后仍有问题，建议尝试使用 paddleOCR 提供的预训练数据集进行测试，以排除数据兼容性问题。

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*

 ⚠️ This is an AI-generated response. Please verify any critical information.
